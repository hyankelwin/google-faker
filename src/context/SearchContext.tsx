import React, { createContext, useState, useCallback, useContext } from 'react';
import { Animal } from '../api/google-faker-api/search/interfaces/animal.interface';
import { getAnimals } from '../api/google-faker-api/search/google-faker.request';
import { getTimeOut } from '../utils/getTimeOutSimulateApi';
import { checkAnimals } from '../utils/getAnimalType';

interface ISearch {
  isLoading: boolean;
  error: boolean;
  search: string;
  animals: Animal[];
  animalSelected: Animal;
  showCard: boolean;
  visibilityModalOptions: boolean;
  visibilityModalSearch: boolean;
  inputIsEmpty: boolean;
  handleFindSeach(value: string): void;
  handleClearInputValue(): void;
  handleSetItemSelected(item: Animal): void;
  handleSetShowCard(): void;
  handleSetClearCard(): void;
  handleSetVisibiltyModalOptions(value: boolean): void;
  handleSetVisibiltyModalSearch(value: boolean): void;
  handleSetEmptyInputValue(value: boolean): void;
}

const SearchContext = createContext<ISearch>({} as ISearch);

const SearchProvider: React.FC = ({ children }) => {
  const [search, setSearch] = useState('');
  const [showCard, setShowCard] = useState(false);
  const [inputIsEmpty, setInputIsEmpty] = useState(true);
  const [visibilityModalOptions, setVisibilityModalOptions] = useState(false);
  const [visibilityModalSearch, setVisibilityModalSearch] = useState(false);
  const [isLoading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  const [animals, setAnimals] = useState<Animal[]>([]);
  const [animalSelected, setAnimalSelected] = useState<Animal>({} as Animal);

  const handleSetItemSelected = useCallback((item: Animal) => {
    setAnimalSelected(item);
  }, []);

  const handleClearInputValue = useCallback(() => {
    setSearch('');
  }, []);

  const handleSetEmptyInputValue = useCallback(value => {
    setInputIsEmpty(value);
  }, []);

  const handleSetShowCard = useCallback(() => {
    setShowCard(true);
  }, []);

  const handleSetClearCard = useCallback(() => {
    setShowCard(false);
  }, []);

  const handleSetLocalStorage = useCallback((value: string) => {
    setShowCard(false);
    localStorage.setItem('search', value);
  }, []);

  const handleSetVisibiltyModalOptions = useCallback((value: boolean) => {
    setVisibilityModalOptions(value);
  }, []);

  const handleSetVisibiltyModalSearch = useCallback((value: boolean) => {
    setVisibilityModalSearch(value);
  }, []);

  const handleFindSeach = useCallback(
    async (value: string) => {
      setLoading(true);
      setError(false);
      setSearch(value);
      handleSetLocalStorage(value);

      await getTimeOut(10);

      if (!checkAnimals(value)) {
        setLoading(false);
        setError(true);
        return;
      }

      const response = await getAnimals(value);

      setAnimals(response);
      setLoading(false);
    },
    [handleSetLocalStorage],
  );

  return (
    <SearchContext.Provider
      value={{
        search,
        isLoading,
        error,
        animals,
        animalSelected,
        showCard,
        visibilityModalOptions,
        visibilityModalSearch,
        inputIsEmpty,
        handleFindSeach,
        handleClearInputValue,
        handleSetItemSelected,
        handleSetShowCard,
        handleSetClearCard,
        handleSetVisibiltyModalOptions,
        handleSetVisibiltyModalSearch,
        handleSetEmptyInputValue,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

function useSearch(): ISearch {
  const context = useContext(SearchContext);

  return context;
}

export { SearchProvider, useSearch };
