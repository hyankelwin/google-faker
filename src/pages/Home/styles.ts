import styled from 'styled-components';

export const Content = styled.div`
  display: grid;
  justify-content: center;
  height: 100vh;
  align-content: center;
  position: relative;
  bottom: 50px;

  @media (max-width: 1300px) {
    justify-content: center;
  }
`;

export const ImageGoogle = styled.img`
  width: 300px;
  background-repeat: no-repeat;
  background-size: cover;
  margin: auto;

  @media (max-width: 600px) {
    width: 200px;
  }

  @media (max-width: 390px) {
    width: 150px;
  }
`;
