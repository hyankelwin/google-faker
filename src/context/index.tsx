import React from 'react';

import { SearchProvider } from './SearchContext';

const AppProvider: React.FC = ({ children }) => (
  <SearchProvider>{children}</SearchProvider>
);

export default AppProvider;
