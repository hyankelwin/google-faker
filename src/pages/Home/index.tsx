import React from 'react';
import { Content, ImageGoogle } from './styles';
import Layout from '../../containers/Layout';
import GoogleImg from '../../assets/images/logo-google.png';
import FormSearch from '../../containers/Form/FormSearch';
import ModalOptions from '../../components/Modals/ModalOptions';
import { useSearch } from '../../context/SearchContext';

const Home: React.FC = () => {
  const { handleSetVisibiltyModalOptions, visibilityModalOptions } =
    useSearch();

  return (
    <Layout page="home">
      <ModalOptions
        isOpen={visibilityModalOptions}
        onCloseModal={() => handleSetVisibiltyModalOptions(false)}
      />
      <Content>
        <ImageGoogle src={GoogleImg} />
        <FormSearch page="home" />
      </Content>
    </Layout>
  );
};

export default Home;
