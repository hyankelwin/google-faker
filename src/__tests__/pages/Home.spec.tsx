import { fireEvent, render, waitFor } from '@testing-library/react';
import { useSearch } from '../../context/SearchContext';
import Home from '../../pages/Home';

jest.mock('../../context/SearchContext');

describe('Home page', () => {
  it('should be able to render on Home page with success', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleSetVisibiltyModalOptions: () => jest.fn(),
      visibilityModalOptions: false,
    });

    const { asFragment } = render(<Home />);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should be able to simulate click close icon modal on Home page', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleSetVisibiltyModalOptions: () => jest.fn(),
      visibilityModalOptions: true,
    });

    const { asFragment, getByTestId } = render(<Home />);

    const iconCloseModal = getByTestId('icon-close-modal');

    fireEvent.click(iconCloseModal);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
