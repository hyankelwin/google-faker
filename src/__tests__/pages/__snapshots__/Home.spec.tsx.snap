// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should be able to render on Home page with success 1`] = `
<DocumentFragment>
  <div
    class="sc-jlZhew dQejNR"
  >
    <div
      class="sc-eqUAAy etcgDI"
    >
      <div>
        <p
          class="sc-gsFSXq jzQfdD"
        >
          Agile Content Frontend test
        </p>
      </div>
      <div>
        <div
          class="sc-iGgWBj iYeObN"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M104 160a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 312a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 464a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56z"
            />
          </svg>
          <img
            alt="Image to simulate user logged"
            class="sc-kAyceB dQecPe"
            src="user.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-dAlyuH NEQFU"
    >
      <div
        class="sc-aXZVg hLSxly"
      >
        <img
          class="sc-gEvEer knjNTS"
          src="logo-google.png"
        />
        <form>
          <div
            class="sc-imWYAI ityMyQ"
          >
            <div
              class="sc-dhKdcB hIcSne"
            >
              <svg
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
              <input
                data-testid="input-search"
                value=""
              />
            </div>
            <div
              class="sc-jXbUNg dwcNzp"
            >
              <button
                class="sc-kpDqfm koXWmj"
                color="rgba(0,0,0,0.87)"
                data-testid="button-submit"
                type="submit"
              >
                Buscar
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="sc-cwHptR kXzOKw"
    >
      <p>
        ©Google 2021
      </p>
      <p>
        Version 0.1.0
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home page should be able to simulate click close icon modal on Home page 1`] = `
<DocumentFragment>
  <div
    class="sc-jlZhew dQejNR"
  >
    <div
      class="sc-eqUAAy etcgDI"
    >
      <div>
        <p
          class="sc-gsFSXq jzQfdD"
        >
          Agile Content Frontend test
        </p>
      </div>
      <div>
        <div
          class="sc-iGgWBj iYeObN"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M104 160a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 312a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 464a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56z"
            />
          </svg>
          <img
            alt="Image to simulate user logged"
            class="sc-kAyceB dQecPe"
            src="user.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-dAlyuH NEQFU"
    >
      <div
        class="sc-aXZVg hLSxly"
      >
        <img
          class="sc-gEvEer knjNTS"
          src="logo-google.png"
        />
        <form>
          <div
            class="sc-imWYAI ityMyQ"
          >
            <div
              class="sc-dhKdcB hIcSne"
            >
              <svg
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
              <input
                data-testid="input-search"
                value=""
              />
            </div>
            <div
              class="sc-jXbUNg dwcNzp"
            >
              <button
                class="sc-kpDqfm koXWmj"
                color="rgba(0,0,0,0.87)"
                data-testid="button-submit"
                type="submit"
              >
                Buscar
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="sc-cwHptR kXzOKw"
    >
      <p>
        ©Google 2021
      </p>
      <p>
        Version 0.1.0
      </p>
    </div>
  </div>
</DocumentFragment>
`;
